// react-native-animated-spinkit/lib/module/Flow.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import * as React from 'react';
import { Animated, Easing, View } from 'react-native';
import { defaultProps } from './SpinnerProps';
import AnimationContainer from './AnimationContainer';
import { stagger } from './utils';
export default class Flow extends React.Component {
  render() {
    const {
      size,
      color,
      style,
      animating,
      hidesWhenStopped,
      ...rest
    } = this.props;
    return /*#__PURE__*/React.createElement(AnimationContainer, {
      initAnimation: () => ({
        flow: value => stagger(150, 3, {
          duration: 1400,
          value: value,
          easing: Easing.bezier(0.455, 0.03, 0.515, 0.955),
          keyframes: [0, 40, 80, 100]
        })
      }),
      animating: animating
    }, values => /*#__PURE__*/React.createElement(View, _extends({
      style: [{
        width: size,
        height: size * 0.25,
        justifyContent: 'space-between',
        flexDirection: 'row',
        opacity: !animating && hidesWhenStopped ? 0 : 1
      }, style]
    }, rest), values.flow.map((value, index) => /*#__PURE__*/React.createElement(Animated.View, {
      key: index,
      style: {
        width: size * 0.25,
        height: size * 0.25,
        borderRadius: size * 0.25 / 2,
        backgroundColor: color,
        transform: [{
          scale: value.interpolate({
            inputRange: [0, 40, 80, 100],
            outputRange: [0.3, 1, 0.3, 0.3]
          })
        }]
      }
    }))));
  }

}

_defineProperty(Flow, "defaultProps", defaultProps);
