// react-native-animated-spinkit/lib/module/Fold.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import * as React from 'react';
import { Animated, Easing, View } from 'react-native';
import { defaultProps } from './SpinnerProps';
import AnimationContainer from './AnimationContainer';
import { stagger } from './utils';
export default class Fold extends React.Component {
  render() {
    const {
      size,
      color,
      style,
      animating,
      hidesWhenStopped,
      ...rest
    } = this.props;
    return /*#__PURE__*/React.createElement(AnimationContainer, {
      initAnimation: () => ({
        fold: value => stagger(300, 4, {
          duration: 2400,
          value: value,
          keyframes: [0, 10, 25, 75, 90, 100],
          easing: Easing.linear
        })
      }),
      animating: animating
    }, values => /*#__PURE__*/React.createElement(View, _extends({
      style: [{
        width: size,
        height: size,
        alignItems: 'center',
        justifyContent: 'center',
        opacity: !animating && hidesWhenStopped ? 0 : 1,
        transform: [{
          rotate: '45deg'
        }]
      }, style]
    }, rest), values.fold.map((value, index) => /*#__PURE__*/React.createElement(Animated.View, {
      renderToHardwareTextureAndroid: true,
      key: index,
      style: {
        width: size / 2,
        height: size / 2,
        backgroundColor: color,
        position: 'absolute',
        opacity: value.interpolate({
          inputRange: [0, 10, 25, 75, 90, 100],
          outputRange: [0, 0, 1, 1, 0, 0]
        }),
        transform: [{
          rotate: "".concat(index * 90, "deg")
        }, {
          perspective: size * 3.6
        }, {
          rotateX: value.interpolate({
            inputRange: [0, 10, 25, 75, 90, 100],
            outputRange: ['-180deg', '-180deg', '0deg', '0deg', '0deg', '0deg']
          })
        }, {
          rotateY: value.interpolate({
            inputRange: [0, 10, 25, 75, 90, 100],
            outputRange: ['0deg', '0deg', '0deg', '0deg', '180deg', '180deg']
          })
        }, {
          translateX: -size / 4
        }, {
          translateY: -size / 4
        }]
      }
    }))));
  }

}

_defineProperty(Fold, "defaultProps", defaultProps);
