// @jsamr/counter-style/lib/cjs/presets/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.upperRoman = exports.upperLatin = exports.upperArmenian = exports.upperAlpha = exports.tibetan = exports.thai = exports.telugu = exports.tamil = exports.square = exports.persian = exports.oriya = exports.myanmar = exports.mongolian = exports.malayalam = exports.lowerRoman = exports.lowerLatin = exports.lowerGreek = exports.lowerArmenian = exports.lowerAlpha = exports.lao = exports.koreanHanjaInformal = exports.koreanHanjaFormal = exports.koreanHangulFormal = exports.khmer = exports.katanaIroha = exports.katana = exports.kannada = exports.japaneseInformal = exports.japaneseFormal = exports.hiraganaIroha = exports.hiragana = exports.hebrew = exports.gurmukhi = exports.gujarati = exports.georgian = exports.disc = exports.devanagari = exports.decimalLeadingZero = exports.decimal = exports.cjkHeavenlyStem = exports.cjkEarthlyBranch = exports.cjkDecimal = exports.circle = exports.cambodian = exports.bengali = exports.armenian = exports.arabicIndic = void 0;
var arabicIndic_1 = require("./arabicIndic");
Object.defineProperty(exports, "arabicIndic", { enumerable: true, get: function () { return __importDefault(arabicIndic_1).default; } });
var armenian_1 = require("./armenian");
Object.defineProperty(exports, "armenian", { enumerable: true, get: function () { return __importDefault(armenian_1).default; } });
var bengali_1 = require("./bengali");
Object.defineProperty(exports, "bengali", { enumerable: true, get: function () { return __importDefault(bengali_1).default; } });
var cambodian_1 = require("./cambodian");
Object.defineProperty(exports, "cambodian", { enumerable: true, get: function () { return __importDefault(cambodian_1).default; } });
var circle_1 = require("./circle");
Object.defineProperty(exports, "circle", { enumerable: true, get: function () { return __importDefault(circle_1).default; } });
var cjkDecimal_1 = require("./cjkDecimal");
Object.defineProperty(exports, "cjkDecimal", { enumerable: true, get: function () { return __importDefault(cjkDecimal_1).default; } });
var cjkEarthlyBranch_1 = require("./cjkEarthlyBranch");
Object.defineProperty(exports, "cjkEarthlyBranch", { enumerable: true, get: function () { return __importDefault(cjkEarthlyBranch_1).default; } });
var cjkHeavenlyStem_1 = require("./cjkHeavenlyStem");
Object.defineProperty(exports, "cjkHeavenlyStem", { enumerable: true, get: function () { return __importDefault(cjkHeavenlyStem_1).default; } });
var decimal_1 = require("./decimal");
Object.defineProperty(exports, "decimal", { enumerable: true, get: function () { return __importDefault(decimal_1).default; } });
var decimalLeadingZero_1 = require("./decimalLeadingZero");
Object.defineProperty(exports, "decimalLeadingZero", { enumerable: true, get: function () { return __importDefault(decimalLeadingZero_1).default; } });
var devanagari_1 = require("./devanagari");
Object.defineProperty(exports, "devanagari", { enumerable: true, get: function () { return __importDefault(devanagari_1).default; } });
var disc_1 = require("./disc");
Object.defineProperty(exports, "disc", { enumerable: true, get: function () { return __importDefault(disc_1).default; } });
var georgian_1 = require("./georgian");
Object.defineProperty(exports, "georgian", { enumerable: true, get: function () { return __importDefault(georgian_1).default; } });
var gujarati_1 = require("./gujarati");
Object.defineProperty(exports, "gujarati", { enumerable: true, get: function () { return __importDefault(gujarati_1).default; } });
var gurmukhi_1 = require("./gurmukhi");
Object.defineProperty(exports, "gurmukhi", { enumerable: true, get: function () { return __importDefault(gurmukhi_1).default; } });
var hebrew_1 = require("./hebrew");
Object.defineProperty(exports, "hebrew", { enumerable: true, get: function () { return __importDefault(hebrew_1).default; } });
var hiragana_1 = require("./hiragana");
Object.defineProperty(exports, "hiragana", { enumerable: true, get: function () { return __importDefault(hiragana_1).default; } });
var hiraganaIroha_1 = require("./hiraganaIroha");
Object.defineProperty(exports, "hiraganaIroha", { enumerable: true, get: function () { return __importDefault(hiraganaIroha_1).default; } });
var japaneseFormal_1 = require("./japaneseFormal");
Object.defineProperty(exports, "japaneseFormal", { enumerable: true, get: function () { return __importDefault(japaneseFormal_1).default; } });
var japaneseInformal_1 = require("./japaneseInformal");
Object.defineProperty(exports, "japaneseInformal", { enumerable: true, get: function () { return __importDefault(japaneseInformal_1).default; } });
var kannada_1 = require("./kannada");
Object.defineProperty(exports, "kannada", { enumerable: true, get: function () { return __importDefault(kannada_1).default; } });
var katana_1 = require("./katana");
Object.defineProperty(exports, "katana", { enumerable: true, get: function () { return __importDefault(katana_1).default; } });
var katanaIroha_1 = require("./katanaIroha");
Object.defineProperty(exports, "katanaIroha", { enumerable: true, get: function () { return __importDefault(katanaIroha_1).default; } });
var khmer_1 = require("./khmer");
Object.defineProperty(exports, "khmer", { enumerable: true, get: function () { return __importDefault(khmer_1).default; } });
var koreanHangulFormal_1 = require("./koreanHangulFormal");
Object.defineProperty(exports, "koreanHangulFormal", { enumerable: true, get: function () { return __importDefault(koreanHangulFormal_1).default; } });
var koreanHanjaFormal_1 = require("./koreanHanjaFormal");
Object.defineProperty(exports, "koreanHanjaFormal", { enumerable: true, get: function () { return __importDefault(koreanHanjaFormal_1).default; } });
var koreanHanjaInformal_1 = require("./koreanHanjaInformal");
Object.defineProperty(exports, "koreanHanjaInformal", { enumerable: true, get: function () { return __importDefault(koreanHanjaInformal_1).default; } });
var lao_1 = require("./lao");
Object.defineProperty(exports, "lao", { enumerable: true, get: function () { return __importDefault(lao_1).default; } });
var lowerAlpha_1 = require("./lowerAlpha");
Object.defineProperty(exports, "lowerAlpha", { enumerable: true, get: function () { return __importDefault(lowerAlpha_1).default; } });
var lowerArmenian_1 = require("./lowerArmenian");
Object.defineProperty(exports, "lowerArmenian", { enumerable: true, get: function () { return __importDefault(lowerArmenian_1).default; } });
var lowerGreek_1 = require("./lowerGreek");
Object.defineProperty(exports, "lowerGreek", { enumerable: true, get: function () { return __importDefault(lowerGreek_1).default; } });
var lowerLatin_1 = require("./lowerLatin");
Object.defineProperty(exports, "lowerLatin", { enumerable: true, get: function () { return __importDefault(lowerLatin_1).default; } });
var lowerRoman_1 = require("./lowerRoman");
Object.defineProperty(exports, "lowerRoman", { enumerable: true, get: function () { return __importDefault(lowerRoman_1).default; } });
var malayalam_1 = require("./malayalam");
Object.defineProperty(exports, "malayalam", { enumerable: true, get: function () { return __importDefault(malayalam_1).default; } });
var mongolian_1 = require("./mongolian");
Object.defineProperty(exports, "mongolian", { enumerable: true, get: function () { return __importDefault(mongolian_1).default; } });
var myanmar_1 = require("./myanmar");
Object.defineProperty(exports, "myanmar", { enumerable: true, get: function () { return __importDefault(myanmar_1).default; } });
var oriya_1 = require("./oriya");
Object.defineProperty(exports, "oriya", { enumerable: true, get: function () { return __importDefault(oriya_1).default; } });
var persian_1 = require("./persian");
Object.defineProperty(exports, "persian", { enumerable: true, get: function () { return __importDefault(persian_1).default; } });
var square_1 = require("./square");
Object.defineProperty(exports, "square", { enumerable: true, get: function () { return __importDefault(square_1).default; } });
var tamil_1 = require("./tamil");
Object.defineProperty(exports, "tamil", { enumerable: true, get: function () { return __importDefault(tamil_1).default; } });
var telugu_1 = require("./telugu");
Object.defineProperty(exports, "telugu", { enumerable: true, get: function () { return __importDefault(telugu_1).default; } });
var thai_1 = require("./thai");
Object.defineProperty(exports, "thai", { enumerable: true, get: function () { return __importDefault(thai_1).default; } });
var tibetan_1 = require("./tibetan");
Object.defineProperty(exports, "tibetan", { enumerable: true, get: function () { return __importDefault(tibetan_1).default; } });
var upperAlpha_1 = require("./upperAlpha");
Object.defineProperty(exports, "upperAlpha", { enumerable: true, get: function () { return __importDefault(upperAlpha_1).default; } });
var upperArmenian_1 = require("./upperArmenian");
Object.defineProperty(exports, "upperArmenian", { enumerable: true, get: function () { return __importDefault(upperArmenian_1).default; } });
var upperLatin_1 = require("./upperLatin");
Object.defineProperty(exports, "upperLatin", { enumerable: true, get: function () { return __importDefault(upperLatin_1).default; } });
var upperRoman_1 = require("./upperRoman");
Object.defineProperty(exports, "upperRoman", { enumerable: true, get: function () { return __importDefault(upperRoman_1).default; } });
