// react-native-animated-spinkit/lib/commonjs/Fold.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _SpinnerProps = require("./SpinnerProps");

var _AnimationContainer = _interopRequireDefault(require("./AnimationContainer"));

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Fold extends React.Component {
  render() {
    const {
      size,
      color,
      style,
      animating,
      hidesWhenStopped,
      ...rest
    } = this.props;
    return /*#__PURE__*/React.createElement(_AnimationContainer.default, {
      initAnimation: () => ({
        fold: value => (0, _utils.stagger)(300, 4, {
          duration: 2400,
          value: value,
          keyframes: [0, 10, 25, 75, 90, 100],
          easing: _reactNative.Easing.linear
        })
      }),
      animating: animating
    }, values => /*#__PURE__*/React.createElement(_reactNative.View, _extends({
      style: [{
        width: size,
        height: size,
        alignItems: 'center',
        justifyContent: 'center',
        opacity: !animating && hidesWhenStopped ? 0 : 1,
        transform: [{
          rotate: '45deg'
        }]
      }, style]
    }, rest), values.fold.map((value, index) => /*#__PURE__*/React.createElement(_reactNative.Animated.View, {
      renderToHardwareTextureAndroid: true,
      key: index,
      style: {
        width: size / 2,
        height: size / 2,
        backgroundColor: color,
        position: 'absolute',
        opacity: value.interpolate({
          inputRange: [0, 10, 25, 75, 90, 100],
          outputRange: [0, 0, 1, 1, 0, 0]
        }),
        transform: [{
          rotate: "".concat(index * 90, "deg")
        }, {
          perspective: size * 3.6
        }, {
          rotateX: value.interpolate({
            inputRange: [0, 10, 25, 75, 90, 100],
            outputRange: ['-180deg', '-180deg', '0deg', '0deg', '0deg', '0deg']
          })
        }, {
          rotateY: value.interpolate({
            inputRange: [0, 10, 25, 75, 90, 100],
            outputRange: ['0deg', '0deg', '0deg', '0deg', '180deg', '180deg']
          })
        }, {
          translateX: -size / 4
        }, {
          translateY: -size / 4
        }]
      }
    }))));
  }

}

exports.default = Fold;

_defineProperty(Fold, "defaultProps", _SpinnerProps.defaultProps);
