// react-native-animated-spinkit/lib/commonjs/Swing.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _SpinnerProps = require("./SpinnerProps");

var _AnimationContainer = _interopRequireDefault(require("./AnimationContainer"));

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Swing extends React.Component {
  render() {
    const {
      size,
      color,
      style,
      animating,
      hidesWhenStopped,
      ...rest
    } = this.props;
    const circleStyle = {
      width: size * 0.45,
      height: size * 0.45,
      backgroundColor: color,
      borderRadius: size * 0.45 / 2
    };
    return /*#__PURE__*/React.createElement(_AnimationContainer.default, {
      initAnimation: () => ({
        swing: value => ({
          values: [value],
          animation: (0, _utils.loop)({
            duration: 1800,
            value: value,
            easing: _reactNative.Easing.linear,
            keyframes: [0, 100]
          })
        }),
        swingDot: value => ({
          values: [value],
          animation: (0, _utils.loop)({
            duration: 2000,
            value: value,
            keyframes: [0, 50, 100]
          })
        })
      }),
      animating: animating
    }, values => /*#__PURE__*/React.createElement(_reactNative.Animated.View, _extends({
      style: [{
        width: size,
        height: size,
        alignItems: 'center',
        justifyContent: 'space-between',
        opacity: !animating && hidesWhenStopped ? 0 : 1,
        transform: [{
          rotate: values.swing[0].interpolate({
            inputRange: [0, 100],
            outputRange: ['0deg', '360deg']
          })
        }]
      }, style]
    }, rest), /*#__PURE__*/React.createElement(_reactNative.Animated.View, {
      style: [circleStyle, {
        transform: [{
          scale: values.swingDot[0].interpolate({
            inputRange: [0, 50, 100],
            outputRange: [0.2, 1, 0.2]
          })
        }]
      }]
    }), /*#__PURE__*/React.createElement(_reactNative.Animated.View, {
      style: [circleStyle, {
        transform: [{
          scale: values.swingDot[0].interpolate({
            inputRange: [0, 50, 100],
            outputRange: [1, 0.2, 1]
          })
        }]
      }]
    })));
  }

}

exports.default = Swing;

_defineProperty(Swing, "defaultProps", _SpinnerProps.defaultProps);
